### ISSUE TITLE
Ability to specify custom error page

### ISSUE THREAD

**Issue by sgringwe** [heptio/contour#320]
It'd be nice to have a way to show a custom error message if an error happens at the contour level, such as no healthy upstream servers.

If there is already a way to do this, it'd be nice to document it.

**Comment by davecheney** [comment]
This sounds like a good thing to have if someone wants to investigate the options inside Envoy for doing this.

**Comment by dobozysaurus** [comment]
Looks like Envoy will support custom error responses in 1.9.0: https://github.com/envoyproxy/envoy/issues/1178.

**Comment by davecheney** [comment]
This is something i’d like to see fo contour 1.0. Marking as 0.15 to review if the feature has landed upstream.

**Comment by stevesloka** [comment]
This feature right now is scheduled for Envoy v1.12.0.

**Comment by davecheney** [comment]
Thanks for checking. I'm moving this to the unplanned milestone. It's unlikely that we'll make it to envoy 1.12 by october for contour 1.0 so we'll reassess after contour 1.0 ships.

**Comment by jpeach** [comment]
xref  #2515 which is closely related

**Comment by anupriya17** [comment]
This will help i think https://github.com/envoyproxy/envoy/pull/11007

**Comment by jpeach** [comment]
Yup, that looks promising.

xref #1898

**Comment by OrlinVasilev** [comment]
@jpeach @stevesloka @sgringwe I think we can close this one right ?

**Comment by stevesloka** [comment]
@OrlinVasilev nope, you'd need a way to configure this in Contour now. =) It's great envoy has support now.

**Comment by OrlinVasilev** [comment]
Great thanks!!

**Comment by oliverkane** [comment]
Hi, just chiming in from what I think was creating a duplicate Issue. Essentially I'm looking for a way to handle a few cases:
- When upstreams are unhealthy, rather than a static 503, I would like to customize the content. This is similar to how some cloud service providers handle things on eg "AWS Application Load Balancers".
- When a wildcard parent route is specified, but the child router is not yet, or invalid, it would be nice to beable to "fall back" to a default handler.

In my days where I'd use the more common NGINX based ingresses, they used a feature called the default backend to accomplish some of these features.

**Comment by Bezbran** [comment]
We're encountering this issue as well.
It's really essential to be able to send custom error pages with HTML/JS/CSS to our users.
For unhealthy upstream errors, 421 errors, 400/404 errors, etc.

**Comment by davinkevin** [comment]
Same requirement

**Comment by tsaarni** [comment]
Envoy has two options
1. `local_reply` ([link](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/local_reply)). It is configured at the HTTP connection manager (listener) level. It can only react to errors generated by Envoy itself (e.g. no healthy upstream, timeouts), not to error responses returned by upstream services. In practice, it cannot be customized per virtualhost / per-route .
2. `custom_response` filter ([link](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/custom_response/v3/custom_response.proto)). It supports per-route configuration. However, it has been marked "work-in-progress" and "not intended for production use", though the status has remained since its introduction in Envoy v1.25.0.

Both are limited to returning a single response body. Additional files like JS/CSS cannot be served with them. `custom_response` additionally supports generating internal redirect to different upstream.

**Comment by Bezbran** [comment]
Hi @tsaarni really appreciate your response.


> Both are limited to returning a single response body. Additional files like JS/CSS cannot be served with them. custom_response additionally supports generating internal redirect to different upstream.

I think we (contour) should support exact the same as envoy. 
We can start with local reply support and having the custom support with feature flag in the Helm chart.
Once we can customize the response body we can add JS/CSS inline just like HTML.
